### PR TITLE
Lfqa demo

### PR DESCRIPTION
**Pull request by seirasto**
# Pull Request      

## What does this PR do?

Add generative capabilities to PrimeQA UI (FiD and FlanT5)

### CODE REVIEW

**Review by hrojek — APPROVED** [review]
LGTM

**Review by kpfadnis — CHANGES_REQUESTED** [review]
@seirasto Left few comments. I see one or two major blockers and few minor fixes.

**Review by bhavani105 — COMMENTED** [review]
Have all 3 readers been tested end to end
Is there REST support for GenerativeReader and PromptReader

File: primeqa/components/reader/generative.py [hunk 1]
```diff
@@ -205,16 +211,14 @@ def predict(
 
         # Run predict
         predictions = {}
+        predictions = {}
```
[the remark this answers is not in the document]
Comment (seirasto) [reply primeqa/components/reader/generative.py:214]: It is twice. Removing one.

File: primeqa/components/reader/prompt.py [hunk 2]
```diff
@@ -271,46 +336,98 @@ class BAMReader(PromptReader):
...
     def train(self, *args, **kwargs):
         pass
 
-    def load(self, *args, **kwargs):
-        self.model = LLMService(token=self.api_key, model_id=self.model_name)
-        self.tokenizer = AutoTokenizer.from_pretrained(self.model_name)
+    def eval(self, *args, **kwargs):
+        pass
+
+    def __hash__(self) -> int:
```
[the remark this answers is not in the document]
Comment (seirasto) [reply primeqa/components/reader/prompt.py:368]: removing duplicate hash in BamReader (rest is good)

File: primeqa/components/reader/prompt.py [hunk 3]
```diff
@@ -1,45 +1,53 @@
...
-            for k, v in self.__class__.__dataclass_fields__.items()
-            if not "exclude_from_hash" in v.metadata
-            or not v.metadata["exclude_from_hash"]
-        ]
-
-        # Step 2: Run
-        return hash(
-            f"{self.__class__.__name__}::{json.dumps({k: v for k, v in vars(self).items() if k in hashable_fields }, sort_keys=True)}"
-        )
+        pass
```
[the remark this answers is not in the document]
Comment (seirasto) [reply primeqa/components/reader/prompt.py:37]: added

File: primeqa/components/reader/prompt.py [hunk 4]
```diff
@@ -271,46 +336,98 @@ class BAMReader(PromptReader):
...
                 top_k=self.top_k,
-                top_p=self.top_p
+                top_p=self.top_p,
             )
-            if "error" in r:
+            if "error" in resp:
                 logger.error("Error running BAM service: ")
-                logger.error(r)
+                logger.error(resp)
                 sys.exit(0)
```
[the remark this answers is not in the document]
Comment (seirasto) [reply primeqa/components/reader/prompt.py:424]: Modified to return None

File: primeqa/services/grpc_server/reader_service.py [hunk 5]
```diff
@@ -147,50 +152,105 @@ def GetAnswers(
...
+                                                ],
+                                            )
+                                            for prediction in predictions_for_context
+                                        ]
+                                    )
+                                    for predictions_for_context in predictions.values()
+                                ]
+                            )
+                        )
+                    else:
```
[the remark this answers is not in the document]
Comment (seirasto) [reply primeqa/services/grpc_server/reader_service.py:210]: This is the current setup that works for everything. All readers other than extractive are generative.

**Review by kpfadnis — APPROVED** [review]
Nice job. LGTM.

**Review by bhavani105 — APPROVED** [review]
LGTM

**Pull request merged**